# SCRIPT

SCRIPT(**S**ingle **C**ell **R**egulatory network **I**nference using ChI**P**-seq Collec**T**ion) is a tool for evaluating the binding enrichment of specific TR at single-cell resolution based on scATAC-seq.

## Installation

Dependency, please install them first:

```bash
libpng12-0 tabix
```

Install SCRIPT

```bash
git clone git@github.com:xindong95/SCRIPT.git
cd SCRIPT
python setup.py install
```

Auto config on *Mercury*(lab in-house server):  

```bash
bash test/config.sh
```

## Usage

### SCRIPT all functions

```log
usage: SCRIPT [-h] [--version] {enrich,setting} ...

SCRIPT

positional arguments:
  {enrich,setting}
    enrich          Main function.
    setting         Configuration.

optional arguments:
  -h, --help        show this help message and exit
  --version         show program's version number and exit

For command line options of each command, type: SCRIPT COMMAND -h
```

### SCRIPT enrich function  

```log
usage: SCRIPT enrich [-h] -e PROCESSED_EXPERIMENT -i FEATURE_MATRIX -s {hs,mm} [-p PROJECT] [--cell_number CELL_NUMBER_IMPUTE]
                     [--peak_confidence PEAK_CONFIDENCE_IMPUTE] [--reference {integration,both,chip,motif}] [-t N_CORES] [-y] [--clean]

optional arguments:
  -h, --help            show this help message and exit

Input files arguments:
  -e PROCESSED_EXPERIMENT, --processed_experiment PROCESSED_EXPERIMENT
                        A processed single cell experiment anndata. Seurat or scanpy or MAESTRO. REQUIRED.
  -i FEATURE_MATRIX, --input_feature_matrix FEATURE_MATRIX
                        A cell by peak matrix . REQUIRED.
  -s {hs,mm}, --species {hs,mm}
                        Species. "hs"(human) or "mm"(mouse). REQUIRED.

Output arguments:
  -p PROJECT, --project PROJECT
                        Project name, which will be used to generate output files folder. DEFAULT: Random generate.

Peak imputation paramater arguments:
  --cell_number CELL_NUMBER_IMPUTE
                        Number of cell to merge together when imputation. DEFAULT: "auto".
  --peak_confidence PEAK_CONFIDENCE_IMPUTE
                        Remove peak that confidence less than this number. (Not including equal to.). Recommand 0.1*cell_number_per_group. DEFAULT:
                        "auto".

Processing options:
  --reference {integration,both,chip,motif}
                        Choise of integration / both / chip / motif, "integration" means integrate motif and ChIP result by standard deviation, "both"
                        means keep the both result without merge. DEFAULT: integration.

Other options:
  -t N_CORES, --thread N_CORES
                        Number of cores use to run SCRIPT. DEFAULT: 16.
  -y, --yes             Whether ask for confirmation. DEFAULT: False.
  --clean               Whether delete tmp files(including bed and search results) generated by SCRIPT. DEFAULT: False.
```

### SCRIPT setting function

```log
usage: SCRIPT setting [-h] [--show] [--human_chip_index HUMAN_CHIP_INDEX] [--human_motif_index HUMAN_MOTIF_INDEX]
                      [--mouse_chip_index MOUSE_CHIP_INDEX] [--mouse_motif_index MOUSE_MOTIF_INDEX]

optional arguments:
  -h, --help            show this help message and exit
  --show
  --human_chip_index HUMAN_CHIP_INDEX
  --human_motif_index HUMAN_MOTIF_INDEX
  --mouse_chip_index MOUSE_CHIP_INDEX
  --mouse_motif_index MOUSE_MOTIF_INDEX
```
